core/src/core_tests/activity_tasks.rs

```diff
@@ -621,3 +623,60 @@ async fn can_heartbeat_acts_during_shutdown() {
     join!(shutdown_fut, complete_fut);
     assert_eq!(&complete_order.into_inner(), &[2, 1])
 }
+
+/// Verifies that if a user has tried to record a heartbeat and then immediately after failed the
+/// activity, that we flush those details before reporting the failure completion.
+#[tokio::test]
+async fn complete_act_with_fail_flushes_heartbeat() {
+    let last_hb = 50;
+    let mut mock_gateway = mock_gateway();
+    let last_seen_payload = Rc::new(RefCell::new(None));
+    let lsp = last_seen_payload.clone();
+    mock_gateway
+        .expect_record_activity_heartbeat()
+        // Two times b/c we always record the first heartbeat, and we'll flush the last
+        .times(2)
+        .returning_st(move |_, payload| {
+            *lsp.borrow_mut() = payload;
+            Ok(RecordActivityTaskHeartbeatResponse {
+                cancel_requested: false,
+            })
+        });
+    mock_gateway
+        .expect_fail_activity_task()
+        .times(1)
+        .returning(|_, _| Ok(RespondActivityTaskFailedResponse::default()));
+
+    let core = mock_core(MocksHolder::from_gateway_with_responses(
+        mock_gateway,
+        [],
+        [PollActivityTaskQueueResponse {
+            task_token: vec![1],
+            activity_id: "act1".to_string(),
+            heartbeat_timeout: Some(Duration::from_millis(1).into()),
```

Comment:
Maybe make this an extremely long value to verify we get throttled before failing the activity.

Reply:
Kinda already happening by asserting the heartbeat RPC is only called twice. Bumped it up anyway in case the test runs slowly